crates/vm/src/eval.rs

```diff
@@ -460,46 +133,63 @@ pub fn build_query(mut result: Box<IterRows>, query: Vec<Query>) -> Result<Box<I
     Ok(result)
 }
 
-/// Optimize & compile the [Expr] for late execution
+/// Optimize & compile the [CrudExpr] for late execution
 #[tracing::instrument(skip_all)]
-pub fn build_ast<P: ProgramVm>(p: &mut P, ast: Expr) -> Result<Code, ErrorVm> {
-    let ast = optimize(p, ast)?;
-    compile(p, ast)
+pub fn build_ast(ast: CrudExpr) -> Code {
+    compile_query_expr(ast.optimize(&|_, _| i64::MAX))
 }
 
-/// Optimize, compile & run the [Expr]
+/// Execute the code
 #[tracing::instrument(skip_all)]
-pub fn run_ast<P: ProgramVm>(p: &mut P, ast: Expr) -> Code {
-    match build_ast(p, ast) {
-        Ok(code) => eval(p, code),
-        Err(err) => Code::Halt(err.into()),
+pub fn eval<P: ProgramVm>(p: &mut P, code: Code) -> Code {
+    match code {
+        Code::Value(_) => code.clone(),
+        Code::Block(lines) => {
+            let mut result = Vec::with_capacity(lines.len());
+            for x in lines {
+                let r = eval(p, x);
+                if r != Code::Pass {
+                    result.push(r);
+                }
+            }
+
+            match result.len() {
+                0 => Code::Pass,
+                1 => result.pop().unwrap(),
+                _ => Code::Block(result),
+            }
+        }
+        Code::Crud(q) => p.eval_query(q).unwrap_or_else(|err| Code::Halt(err.into())),
+        Code::Pass => Code::Pass,
+        Code::Halt(_) => code,
+        Code::Table(_) => code,
     }
 }
 
-// Used internally for testing recursion
-#[doc(hidden)]
-pub fn fibo(input: u64) -> Expr {
-    let ty = AlgebraicType::U64;
-
-    let less = |val: u64| bin_op(OpMath::Minus, var("n"), scalar(val));
-
-    let f = Function::new(
-        "fib",
-        &[Param::new("n", ty.clone())],
-        ty,
-        &[if_(
-            bin_op(OpCmp::Lt, var("n"), scalar(2u64)),
-            var("n"),
-            bin_op(
-                OpMath::Add,
-                call_fn("fib", &[("n", less(1))]),
-                call_fn("fib", &[("n", less(2))]),
-            ),
-        )],
-    );
-    let f = Expr::Fun(f);
-
-    Expr::Block(vec![f, call_fn("fib", &[("n", scalar(input))])])
+fn to_vec(of: Vec<Expr>) -> Code {
+    let mut new = Vec::with_capacity(of.len());
+    for ast in of {
+        let code = match ast {
+            Expr::Block(x) => to_vec(x),
+            Expr::Crud(x) => build_ast(*x),
+            x => Code::Halt(ErrorVm::Unsupported(format!("{x:?}")).into()),
+        };
+        new.push(code);
+    }
+    Code::Block(new)
+}
+
+/// Optimize, compile & run the [Expr]
+#[tracing::instrument(skip_all)]
+pub fn run_ast<P: ProgramVm>(p: &mut P, ast: Expr) -> Code {
+    let code = match ast {
+        Expr::Block(x) => to_vec(x),
+        Expr::Crud(x) => build_ast(*x),
+        Expr::Value(x) => Code::Value(x),
+        Expr::Halt(err) => Code::Halt(err),
+        Expr::Ident(x) => Code::Halt(ErrorVm::Unsupported(format!("Ident {x}")).into()),
```

Comment:
Do we still need `Expr::Ident`?

Reply:
Yes, is not executed here but the upper layer building the `SELECT column_name <- ident` need it.